lib/hanami/application/router.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+require "hanami/router"
+require "rack"
+
+module Hanami
+  class Application
+    # Hanami application router
+    # @since 2.0.0
+    class Router < Hanami::Router
+      def initialize(**options, &block)
+        @options = options
+        @middlewares = []
+        super
+        @rack_app = build_rack_app
+      end
+
+      def use(*args, &block)
+        @middlewares << (args << block)
+      end
+
+      # rubocop:disable Metrics/MethodLength
+      def mount(app, at:, host: nil, &block)
+        if app.is_a?(Symbol)
+          # TODO: I wonder if this is the actual behaviour we want...
+          raise "Slices can only be mounted from top-level of routes" unless @context.respond_to?(:slices)
+
+          # TODO: store slices in a way that's more efficient for lookup
+          slice = @context.slices[app]
+
+          # TODO: real exception class
+          raise "Slice +#{app}+ not found" unless slice
+
+          slice_router = self.class.new(
+            **@options,
+            context: slice,
+            endpoint_resolver: @endpoint_resolver.with_container(slice),
+            &block
+          )
+
+          super(slice_router, at: at, host: host)
+        else
+          super(app, at: at, host: host)
+        end
+      end
+      # rubocop:enable Metrics/MethodLength
+
+      alias call_router call
+      private :call_router # rubocop:disable Style/AccessModifierDeclarations
+
+      def call(env)
+        # FIXME: The to_a is here to ensure we return a proper rack-conformant
+        # 3-element array (which our rack_monitor middleware needs), because
+        # sometimes we get an Hanami::Action::Response object instead. This
+        # should not really be necessary.
+        @rack_app.call(env).to_a
+      end
+
+      private
+
+      def build_rack_app
+        middlewares = @middlewares
```

Comment:
The idea for the future: can we make specific order for middlewares? It will be useful when we need to setup auth before all but after `request id` middleware or something like this. WDYT about it?